artifactory.py

```diff
@@ -978,6 +978,49 @@ def get_response(self, pathobj):
 
         return response
 
+    def deploy_by_checksum(
```

Comment:
I do not think we need a separate method for deployment. Most of it is already covered by `deploy()`

from JFrog docs: `If the X-Checksum-Deploy header is set to false, the artifact will be uploaded successfully with a 201 response, even if it didn't exist before, and submitted checksums will have status Uploaded: Identical`

so if we provide argument `by_checksum` we assume that this header should be used, all other parameters seems to be almost the same

Reply:
If there are more big artifacts that need uploading, the `deploy_by_checksum` function would save more network flow and time than `deploy`. Is this slightly better than `deploy`? The key difference is `deploy` uses `rest_put_stream` while `deploy_by_checksum` uses `rest_put`.

`If Artifactory already contains a user readable artifact with the same checksum the artifact content is copied over to the new location and returns a response without requiring content transfer.`

Reply:
I do not think stream parameter automatically uploads all the content to the library. This is decided by headers. 

I think one can use deploy function as it right now and add there new parameter `by_checksum` that will send like a copy request to find objects by checksum. If `by_checksum`  set to `False`, then you just do normal upload of the artifact

from JFrog API docs all deployment functions look similar and different only in headers, then why should we create new function which is 80% identical

@allburov 
you thoughts?
I recommend to use single `deploy` method for everything

Reply:
Could we test it out - if `deploy` with [the right configured headers](https://www.jfrog.com/confluence/display/JFROG/Artifactory+REST+API#ArtifactoryRESTAPI-DeployArtifactbyChecksum) works the same way as `deploy_by_checksum` without uploading the data to Artifactory Server let's use this then/

Reply:
Hi @allburov 
Thanks for your attention.

Please see red items on attached screen shot..

Even `X-Checksum-Deploy=true`, `X-Checksum-Sha1` and `X-Checksum-Sha256` are set in headers, the request body still contains file data.

That's to say, more network flow and time are required if `deploy` with right configured headers is called.

<img width="1792" alt="image" src="https://user-images.githubusercontent.com/32819197/131783176-9ddd8e9d-c7fb-49cd-b773-2885d1747a31.png">

Reply:
:param data: (optional) Dictionary, list of tuples, bytes, or **_file-like object to send in the body_** of the :class:`Request`.

Reference: [requests.put](https://docs.python-requests.org/en/latest/_modules/requests/api/#put)

Reply:
@zhan9san 
OK, this is fine to have `deploy_by_checksum` as high level function in ArtifactoryPath
why do you need to redefine deploy in accessor? you just set `fobj` to None if header for deploying by checksum is enabled

Reply:
Too many switches are not good code-style, IMHO.

Two methods are fine for me.

@zhan9san could you provide an example of combination of two functions while them in README.md? I'm not sure I got the idea behind `deploy_by_checksum`, what the next step if we got an error due to there's no file with the same hash? Will we  upload the original content via `deploy` then?

I'm thinking about combining them together, like "Try to deploy via hash and if it failed - upload a whole file to Artifactory" function. It may be helpful

Reply:
@allburov 
if you compare code of accessor deploy, you will see that it is identical, except headers
as I wrote

> this is fine to have `deploy_by_checksum` as high level function in ArtifactoryPath
> why do you need to redefine deploy in accessor? you just set `fobj` to None if header for deploying by checksum is enabled

otherwise it is a repetition of the code





> I'm thinking about combining them together, like "Try to deploy via hash and if it failed - upload a whole file to Artifactory" function. It may be helpful

this sounds dangerous. May be user excepts fail and does not want to upload, here is JFrog doc:
_If Artifactory already contains a user readable artifact with the same checksum the artifact content is copied over to the new location and returns a response without requiring content transfer.
Otherwise, a 404 error is returned to indicate that content upload is expected in order to deploy the artifact._

Reply:
It's fine for me to use `deploy` in accessor. I'll remove `deploy_by_checksum` in accessor. @allburov  Do you have concern?

> combining them together

I agree with @beliaev-maksim . `deploy_by_sum` is a wrapper of real API. It's end user's responsibility to catch 404 error.